stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,372 @@
+    from stardist.models import StarDist2D, StarDist3D
+    isinstance(model, (StarDist2D, StarDist3D)) or _raise(ValueError("not a valid model"))
+    0 <= min_percentile < max_percentile <= 100 or _raise(ValueError("invalid percentile values"))
+
+    outpath = Path(outpath)
+    if outpath.suffix == "":
+        outdir = outpath
+        zip_path = outdir / f"{name}.zip"
+    elif outpath.suffix == ".zip":
+        outdir = outpath.parent
+        zip_path = outpath
+    else:
+        raise ValueError(f"outpath has to be a folder or zip file, got {outpath}")
+    outdir.mkdir(exist_ok=True, parents=True)
+
+    kwargs = _get_stardist_metadata(outdir)
+    model_kwargs = _get_weights_and_model_metadata(outdir, model, test_input, test_input_axes, test_input_norm_axes, mode,
+                                                   min_percentile=min_percentile, max_percentile=max_percentile)
+    kwargs.update(model_kwargs)
+    kwargs.update(overwrite_spec_kwargs)
+
+    build_model(name=name, output_path=zip_path, **kwargs)
+
+
+class BioimageioModel():
+    def __init__(self, rdf):
+        """Load bioimage.io StarDist model (2D or 3D) from an RDF string (DOI, URL, or file path).
+
+        Note: This is currently not as full-featured as a standard StarDist2D or StarDist3D model.
+        """
+        *_, self.bioimageio_core, self.xr = _import()
+        from bioimageio.core.prediction_pipeline import create_prediction_pipeline
+        self.model = self.bioimageio_core.load_resource_description(rdf)
+        self.pipeline = create_prediction_pipeline(bioimageio_model=self.model)
+        self.thresholds = self.model.config['stardist'].get('thresholds', dict(prob=0.5,nms=0.4))
+        assert self.model.config['stardist']['config']['n_dim'] in (2,3)
+        from stardist.models import Config2D, Config3D
+        self.config = (Config2D if self.model.config['stardist']['config']['n_dim'] == 2 else Config3D)(**self.model.config['stardist']['config'])
+
+
+    def predict(self, img, axes=None):
+        """Predict.
+
+        Parameters
+        ----------
+        img : :class:`numpy.ndarray`
+            Input image
+        axes : str or None
+            Axes of the input ``img``.
+            ``None`` denotes that axes of img are the same as those of the model.
+
+        Returns
+        -------
+        (:class:`numpy.ndarray`, :class:`numpy.ndarray`)
+            Returns the tuple (`prob`, `dist`) of per-pixel object probabilities and star-convex polygon/polyhedra distances.
+
+        """
+        axes_net = self.model.inputs[0].axes
+        if axes is None:
+            axes = ''.join(axes_net).replace('b','')
+            if img.ndim == len(axes)-1 and self.config.n_channel_in == 1:
+                axes = axes.replace('c','')
+        img = move_image_axes(img, axes, ''.join(axes_net).replace('b','s'), adjust_singletons=True)
+        x = self.xr.DataArray(img, dims=axes_net)
+        y = self.pipeline(x)[0]
```

Comment:
Note that prediction pipeline does not do any padding or tiling, for this you would need to use [predict_with_padding](https://github.com/bioimage-io/core-bioimage-io-python/blob/main/bioimageio/core/prediction.py#L253) / [predict_with_tiling](https://github.com/bioimage-io/core-bioimage-io-python/blob/main/bioimageio/core/prediction.py#L274).

Reply:
I don't really understand.

If I want to predict with an input image that has an incompatible size (e.g. not divisible by 8 for spatial dimensions), do I then have to use `predict_with_padding`?

And `predict_with_tiling` has no padding option, because the image is broken up into tiles anyway? (And tile shapes are assumed to be compatible with the neural network?)

Reply:
> If I want to predict with an input image that has an incompatible size (e.g. not divisible by 8 for spatial dimensions), do I then have to use predict_with_padding?

Yes, for prediction with an incompatible shape, but where you don't want to tile use `predict_with_padding` and pass `padding={"x": 16, "y": 16, z": 4}`, where the numbers are the mininimal divisor for the corresponding axis. (I will also work on support passing `padding=True`, so that this is directly read from the spec.)
Note that the returned prediction will always have the same size as the input; it will be padded internally before prediction and then the padding will be cropped again before returning the output.

> And predict_with_tiling has no padding option, because the image is broken up into tiles anyway? (And tile shapes are assumed to be compatible with the neural network?)

Exactly, you need to pass `tiling={"tile": {"x": 256, "y": 256}, "halo": {"x": 32, "y": 32}}`, where `"tile"` is the (inner) tile size and `"halo"` the overlap between tiles. The `tile + 2 * halo` must be a valid input for the network, because that's the size of the input for each tile. The tiling function will make sure internally that all tiles have this size, i.e. it will pad border tiles to the correct shape.
Again, the output will be cropped internally so that it has the same spatial size as the input.
(I will also work in supporting `tiling=True` for getting the parameters from the spec).

I hope that clarifies things.

I also realized that this is not documented for the functions themselves at all, but only documented somewhere else. I will add function signatures here as well.